server/sonar-server/src/main/java/org/sonar/server/es/queue/RecoveryIndexer.java

```diff
@@ -0,0 +1,137 @@
+
+  private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1,
+    new ThreadFactoryBuilder()
+      .setPriority(Thread.MIN_PRIORITY)
+      .setNameFormat("RecoveryIndexer-%d")
+      .build());
+  private final System2 system2;
+  private final Settings settings;
+  private final DbClient dbClient;
+  private final UserIndexer userIndexer;
+
+  // initialized at startup
+  private long minAgeInSeconds;
+  private long maxAgeInSeconds;
+  private long maxRecoveryDurationInSeconds;
+
+  public RecoveryIndexer(System2 system2, Settings settings, DbClient dbClient, UserIndexer userIndexer) {
+    this.system2 = system2;
+    this.settings = settings;
+    this.dbClient = dbClient;
+    this.userIndexer = userIndexer;
+  }
+
+  @Override
+  public void start() {
+    long delayInSeconds = getSetting(PROPERTY_DELAY, DEFAULT_DELAY_IN_SECONDS);
+    minAgeInSeconds = getSetting(PROPERTY_MIN_AGE, DEFAULT_MIN_AGE_IN_SECONDS);
+    maxAgeInSeconds = getSetting(PROPERTY_MAX_AGE, DEFAULT_MAX_AGE_IN_SECONDS);
+    maxRecoveryDurationInSeconds = getSetting(PROPERTY_MAX_RECOVERY_DURATION, DEFAULT_MAX_RECOVERY_DURATION_IN_SECONDS);
+    checkState(minAgeInSeconds < maxAgeInSeconds,
+      "Parameter [%s: %s] must be lower than [%s: %s]", PROPERTY_MAX_AGE, minAgeInSeconds, PROPERTY_MAX_AGE, maxAgeInSeconds);
+
+    executorService.scheduleAtFixedRate(
+      this::index,
+      0,
+      delayInSeconds,
+      TimeUnit.SECONDS);
+  }
+
+  @Override
+  public void stop() {
+    executorService.shutdown();
+  }
+
+  protected void index() {
+    long startedAt = system2.now();
+    long beforeDate = startedAt - minAgeInSeconds * 1_000;
+    long afterDate = startedAt - maxAgeInSeconds * 1_000;
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Collection<EsQueueDto> items = dbClient.esQueueDao().selectForRecovery(dbSession, beforeDate, afterDate);
+      while (!items.isEmpty()) {
+        if (system2.now() > startedAt + maxRecoveryDurationInSeconds * 1_000) {
+          LOGGER.warn("Elasticsearch recovery - timeout for recovery reached. Documents will be indexed at next run (see property {})", PROPERTY_DELAY);
+          break;
+        }
+        ListMultimap<EsQueueDto.Type, EsQueueDto> itemsByType = items.stream()
+          .collect(MoreCollectors.index(EsQueueDto::getDocType));
+        itemsByType.asMap().forEach((type, typeItems) -> {
+          LOGGER.trace("Elasticsearch recovery - processing {} {}", typeItems.size(), type);
+          switch (type) {
+            case USER:
+              userIndexer.index(dbSession, typeItems);
+              break;
+            default:
+              LOGGER.error("Elasticsearch recovery - ignore {} documents with unsupported type {}", typeItems.size(), type);
+              break;
+          }
+        });
+        items = dbClient.esQueueDao().selectForRecovery(dbSession, beforeDate, afterDate);
+      }
+    } catch (Throwable t) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1181)